client/finality-grandpa/src/authorities.rs

```diff
@@ -152,12 +167,45 @@ pub struct AuthoritySet<H, N> {
 	/// is lower than the last finalized block (as signaled in the forced
 	/// change) must be applied beforehand.
 	pending_forced_changes: Vec<PendingChange<H, N>>,
+	/// Track at which blocks the set id changed. This is useful when we need to prove finality for a
+	/// given block since we can figure out what set the block belongs to and when the set
+	/// started/ended.
+	authority_set_changes: AuthoritySetChanges<N>,
+}
+
+impl<H, N> Decode for AuthoritySet<H, N>
+where
+	H: Decode,
+	N: Decode + Clone + Ord
+{
+	fn decode<I: Input>(value: &mut I) -> Result<Self, parity_scale_codec::Error> {
+		let legacy = LegacyAuthoritySet::decode(value)?;
+		let authority_set_changes = match <AuthoritySetChanges<N>>::decode(value) {
+			Ok(v) => v,
+			Err(_) => AuthoritySetChanges::empty(),
+		};
```

Comment:
Where is this decode being used @andresilva ?

If someone uses this where there is still data in `I`, this could lead to decoding shit :P

Reply:
Currently it's only used through this https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/aux_schema.rs#L110, so I don't think it is ever decoded as part of a bigger structure (i.e. there is no more data to decode). It would be nice that we don't have a footgun here though, do you have any suggestion?

Reply:
Maybe we should offer some migration? So, that we only apply this once? But I assume we don't have any way to store a version in aux?

Reply:
Besides that, I don't have better ideas.

Reply:
I was already not remembering the `aux_schema` code for GRANDPA. We indeed do have a version there and we have code there that handles migrations of GRANDPA data. So yeah, we should definitely handle this the same way.